### DISCUSSION TITLE
Problem with Custom Validator  after update to v4.1.0

### DISCUSSION THREAD

**Discussion by PeterKnotek** [netbox-community/netbox#17377] [General]
Hi 

Have problem with cutom validator after update Netbox to 4.1

Here is CODE 
```
from django.db.models import Q
from extras.validators import CustomValidator
from dcim.models import Device
from tenancy.models import Tenant
from django.core.exceptions import ValidationError
from django.contrib.auth import get_user_model
from django.core.exceptions import PermissionDenied
from django.urls import reverse


import re

class MyValidatorHelper:

## This validator checks if there is a Tenant with the same name and allows saving the TenantName with only allowed characters ##
    @staticmethod
    def duplicateValidator(needle, haystack):
        fail_message = []
        for hay in haystack:
            if needle.id != hay.id:
                if needle.name.lower() == hay.name.lower():
                    fail_message.append(" Tetant s nazvom {} sa pouziva !!".format(hay.name))
        return fail_message


class TenantNameValidator(CustomValidator):

    def validate(self, instance):
        fail_message = []
        # Format
        if not re.match(r'^[A-Z0-9_.-]*$', instance.name):
            fail_message.append(" ## NotAllowedCharacter ## !! AllowedCharacter == A-Z | -_ and . )")
        # DuplicateName
        fail_message += MyValidatorHelper.duplicateValidator(instance, Tenant.objects.all())
        if fail_message:
            self.fail("Error !!! : {}".format(" and ".join(fail_message)), field='name')
```

Netbox write
CstomXYZValidator.validate() takes 2 positional arguments but 3 were given

Can help me ???
Thanks

**Comment by candlerb** [comment]
> Have problem with cutom validator after update Netbox to 4.1

Can you describe the problem?  What do you do to trigger the problem, what happens, and what did you expect to happen instead?  Does it show an exception?

Aside: if you could edit your post and insert three backticks (` ``` `) on a line of their own, before and after your code, it will format properly.

**Reply by candlerb** [reply]
You added:

> CstomXYZValidator.validate() takes 2 positional arguments but 3 were given

That's pretty clear. See the [changelog for v4.1.0](https://github.com/netbox-community/netbox/releases/v4.1.0) under "Breaking changes":

* The `validate()` method on CustomValidator subclasses now **must** accept the request argument (deprecated in v4.0 by https://github.com/netbox-community/netbox/issues/14279/).

See also example validation function [here](https://netboxlabs.com/docs/netbox/en/stable/customization/custom-validation/#custom-validation-logic).

**Comment by PeterKnotek** [comment]
Thanks all

Must edit one LINE
From :
 def validate(self, instance):
To:
def validate(self, instance, *args, **kwargs):

**Reply by candlerb** [reply]
To:

```
def validate(self, instance, request):
```

or if you want it to work on pre-4.0:

```
def validate(self, instance, request=None):
```

**Comment by PeterKnotek** [comment]
I have a question

Can I define in the custom validator - for field XYZ or custom field
-  can only be updated by user XYZ???
- for example only admin ??

Thanks

**Reply by candlerb** [reply]
Yes, the username is available as part of the `request` parameter. See [here](https://netboxlabs.com/docs/netbox/en/stable/customization/custom-validation/#validating-request-parameters).
